### PR TITLE
feat: build ARM-architecture-based images as AMD64

### LINKED ISSUE #1949: Support building on ARM Macs
Today, when customers deploy their service, copilot builds the container image and pushes it. Customers using ARM macs end up having ARM based images built and pushed. These images cannot run on Fargate and lead to a confusing and poor customer experience. I propose:

1. First, bailing out early if the image architecture cannot run on Fargate 
2. Support building the image according to the target architecture. Currently Fargate only supports x86 so we should target that (or AMD64).

https://github.com/aws/copilot-cli/issues/1948

https://twitter.com/elliottkember/status/1361169248488316932?s=21

### PR DESCRIPTION
**Pull request by huanjani**
Previously, users could not build ARM-based images with Copilot (without employing workarounds) because Fargate does not currently support ARM-based tasks. This PR automatically builds images as `linux/amd64` when ARM (or other non-AMD) architectures are detected, using Docker's `--platform` flag. This is a best-effort redirect, as users' Docker servers must be multi-platform-capable in order for `--platform` to work. A warning is logged at `svc init` and info is logged at `svc deploy` IF non-amd64 architectures are detected.

An additional manifest field, `platform:` is surfaced iff OS/arch is being redirected, to provide visibility to the user and for extensibility (we can take changes in the future). Validation is performed when the manifest is unmarshaled. The field currently takes a string, but will accept a map in a future PR. 

_The workflow/experience does not change for amd64 users!_

Related: #2545; closes: #1949, #2369.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by huanjani** [comment]
Again, the experience remains the same for amd64 machines.
This is what an ARM-based user sees:
```
$ copilot svc init
Note: It's best to run this command in the root of your workspace.
Service type: Load Balanced Web Service
Service name: svc
Dockerfile: ./Dockerfile
Note: Architecture type arm64 is currently unsupported. Setting platform linux/amd64 instead.
```
manifest:
```
# Your service name will be used in naming your resources like log groups, ECS services, etc.
name: svc
type: Load Balanced Web Service
platform: linux/amd64
# Distribute traffic to your service.
```
```
$ copilot svc deploy
Service name: svc
Only found one environment, defaulting to: test
✔ Upgraded environment test's template to version v1.5.0.on v1.5.0. 
Building your container image: docker build -t 24059xxxxxxx.dkr.ecr.us-east-1.amazonaws.com/demo-app/svc --platform linux/amd64 /home/ec2-user/e
csdemo-frontend -f /home/ec2-user/ecsdemo-frontend/Dockerfile
[+] Building 0.6s (4/10)                                                                                                                        
[+] Building 0.7s (5/10)                                                                                                                        
 => [internal] load build definition from Dockerfile                                                                                       0.0s.........
```

File: templates/workloads/services/lb-web/manifest.yml
Comment (efekarakus) [inline templates/workloads/services/lb-web/manifest.yml]: can we move this block below `memory` in all manifests? 🙏 
I see this as a task-level configuration so we can group them together.

File: internal/pkg/manifest/workload.go
Comment (efekarakus) [inline internal/pkg/manifest/workload.go]: Is it possible to define `Platform` in a single location? preferably in `TaskConfig` 🙏 

This will allow users to define `platform` under `image.build` because of the `yaml` tags which I think we are trying to avoid 🤔 

Maybe we can do it through 💭 
```go
func (s *BackendService) TaskPlatform() string
func (s *BackendService) BuildArgs() *DockerBuildArgs
```

File: internal/pkg/cli/svc_init.go
Comment (efekarakus) [inline internal/pkg/cli/svc_init.go]:
```suggestion
	os, arch := runtime.GOOS, runtime.GOARCH
```

File: internal/pkg/cli/svc_init.go
Comment (efekarakus) [inline internal/pkg/cli/svc_init.go]: do we not need to set os to `linux`? won't this be `darwin` for mac users

**Review by efekarakus — COMMENTED** [review]
Feel free to not be blocked by my comments! I just took a quick glance :)

File: internal/pkg/cli/svc_init.go
Comment (efekarakus) [inline internal/pkg/cli/svc_init.go]: I think it shouldn't be the responsibility of this function to check if `image` is empty, can we instead merge the two functions into only `redirectPlatform`:
```go
if image == "" {
   return "", nil // Nothing to do.
}
os, arch, err := engine.GetPlatform()
if err != nil { ... }

if arch != exec.Amd64Arch { ... }
return "", nil
```

File: internal/pkg/cli/svc_deploy.go
Comment (efekarakus) [inline internal/pkg/cli/svc_deploy.go]: Should this variable be defined in the `exec` pkg? where the other constants are

File: internal/pkg/cli/svc_deploy.go
Comment (efekarakus) [inline internal/pkg/cli/svc_deploy.go]: similarly can we move this to the `exec` pkg? it'd be awesome if we can move all this functionality to a completely separate `docker` or `docker/engine` pkg

File: internal/pkg/manifest/workload.go
Comment (efekarakus) [inline internal/pkg/manifest/workload.go]: nit: comments can now be removed :D

File: internal/pkg/cli/svc_deploy.go
Comment (huanjani) [reply internal/pkg/cli/svc_deploy.go]: Yes! And I have a PR waiting in the wings (for after this one is merged) that splits `docker` into `dockerengine` and `dockerfile`.

File: internal/pkg/cli/svc_init.go
Comment (huanjani) [reply internal/pkg/cli/svc_init.go]: Wait, you mean `if image != ""`, right? 🤔 
That's the case in which we wouldn't build so the `--platform` flag is irrelevant.

Comment (efekarakus) [reply internal/pkg/cli/svc_init.go]: Yea sorry 😅

File: internal/pkg/exec/docker.go
Comment (Lou1415926) [inline internal/pkg/exec/docker.go:115]: User friendly!!

File: internal/pkg/exec/docker.go
Comment (Lou1415926) [inline internal/pkg/exec/docker.go]: 🙋‍♀️ why do we prefer checking for each individual parts instead of the whole string `plaform`?

File: internal/pkg/exec/docker.go
Comment (Lou1415926) [inline internal/pkg/exec/docker.go]: We should take advantage of this package https://github.com/dustin/go-humanize here!

For example,
```suggestion
		return fmt.Errorf("platform %s is invalid; %s: %s", 
		english.PluralWord(len(ValidPlatforms), "the valid platform is", "valid platforms are"),
		english.WorkSeries(ValidPlatforms, "and"))
```

File: internal/pkg/manifest/job.go
Comment (Lou1415926) [inline internal/pkg/manifest/job.go]: Perhaps we can take advantage of go's composition and define a function with `TaskConfig` being the receiver:
```suggestion
func (t *TaskConfig) TaskPlatform() string {
  return t.Platform
}
```
This way we don't need to maintain several functions for a single purpose - same with `AppRunnerInstanceConfig`!

File: internal/pkg/cli/svc_init.go
Comment (Lou1415926) [inline internal/pkg/cli/svc_init.go]: Do you think this function would be better suited in the `docker` package?

File: templates/workloads/services/worker/manifest.yml
Comment (efekarakus) [inline templates/workloads/services/worker/manifest.yml:31]: Can we also add documentation for this feature 🙏 ?

I think we can just say that we 'll pass whatever value is written here to `docker build --platform` for now.

File: templates/workloads/services/worker/manifest.yml
Comment (Lou1415926) [inline templates/workloads/services/worker/manifest.yml:30]: Will this be evaluated to true if `.Platform == ""`?

File: internal/pkg/manifest/job.go
Comment (Lou1415926) [inline internal/pkg/manifest/job.go]: `t.Platform` is a user input read from manifest right 🤔 ? I may be missing something from previous PRs - we don't do anything detecting after reading in the manifest right?

File: internal/pkg/exec/docker.go
Comment (efekarakus) [inline internal/pkg/exec/docker.go]: can this be private?

File: internal/pkg/cli/svc_init.go
Comment (Lou1415926) [inline internal/pkg/cli/svc_init.go]: We can probably also direct people's attention to the manifest's `platform` field in this case!

File: internal/pkg/cli/svc_deploy.go
Comment (Lou1415926) [inline internal/pkg/cli/svc_deploy.go]: Do we also need these lines for `job deploy`?

Also, do you think the validation function could be invoked from `TaskPlatform`? So the function signature would be `func (t *TaskConfig) TaskPlatform() (string, error)` which returns an error if the platform is invalid.

**Review by Lou1415926 — COMMENTED** [review]
yayy this would fill the blank in our current user experience regarding platform validity!!

A few thoughts and questions ⬇️ !!

File: internal/pkg/exec/docker.go
Comment (efekarakus) [inline internal/pkg/exec/docker.go]: It is slightly strange to export a format string :D 

What do you think of instead making this private and creating a function instead?
```go
func DockerBuildPlatform(os, arch string)  string { return fmt.Sprintf("%s/%s", os, arch) }
```

File: internal/pkg/cli/svc_init.go
Comment (efekarakus) [inline internal/pkg/cli/svc_init.go]: nit: should we name this `platform` to keep it consistent?

**Review by efekarakus — COMMENTED** [review]
Yay looks great! Feel free to ignore the nits, just minor comments

File: internal/pkg/exec/docker.go
Comment (huanjani) [reply internal/pkg/exec/docker.go]: Yes. I had to make it public when I moved it into the `exec` package, but I can just hard code the value(s) in the unit tests.

Comment (efekarakus) [reply internal/pkg/exec/docker.go]: No worries! It can remain public then 👍

File: internal/pkg/exec/docker.go
Comment (huanjani) [reply internal/pkg/exec/docker.go]: Haha, yes, I originally had a "stringifyPlatform" function that I did away with for the const. Will switch back to function.

File: internal/pkg/exec/docker.go
Comment (huanjani) [reply internal/pkg/exec/docker.go]: Great question! I had written this validation before adding the `validPlatforms` var, so breaking them up seemed to avoid a laundry list of combinations in the future, but you're right-- we should just validate the platform as a whole. Changing!

File: internal/pkg/manifest/job.go
Comment (huanjani) [reply internal/pkg/manifest/job.go]: So good! Thanks!

File: internal/pkg/manifest/job.go
Comment (huanjani) [reply internal/pkg/manifest/job.go]: Yes. Changed the comment.

File: internal/pkg/cli/svc_deploy.go
Comment (huanjani) [reply internal/pkg/cli/svc_deploy.go]: `job_deploy` uses the `buildArgs()` function in `svc_deploy`, but I'll manually test to double check.

I stuck the validation in `TaskPlatform()`! 👍🏼

File: templates/workloads/services/worker/manifest.yml
Comment (huanjani) [reply templates/workloads/services/worker/manifest.yml:30]: nope!

File: internal/pkg/cli/svc_init.go
Comment (huanjani) [reply internal/pkg/cli/svc_init.go]: 🤔  I can see it both ways, but sure-- makes sense, moving!

**Review by efekarakus — APPROVED** [review]

File: internal/pkg/manifest/rd_web_svc.go
Comment (Lou1415926) [inline internal/pkg/manifest/rd_web_svc.go]: I worry that having `Platform` as `string` instead of `*string` would result in it getting overriden with `""` if the environment manifest config doesn't specify `platform` 🤔 . 

If it gets overridden as `""`, that could be a problem for users that are not using `linux/amd64` right? If this indeed is a potential impact, could we double check in `ApplyEnv`'s unit test?

Same with the `Platform` in `TaskConfig`.

Sorry for not seeing this in my previous review!!! 🙏🏼

**Review by Lou1415926 — COMMENTED** [review]
Just one more comment on manifest 🙏🏼 !!

File: internal/pkg/exec/docker.go
Comment (Lou1415926) [inline internal/pkg/exec/docker.go:49]: I'm wondering if we should it as `string` here to keep it consistent with the other `string` fields such as `Target` 🤔?

**Review by Lou1415926 — APPROVED** [review]
Let's :shipit:

**Pull request merged**
